### PR TITLE
feat: support external options

### LINKED ISSUE #3615: Support options in external plugins
#3433 moved the definition of options to our internal parsers but there's still a bit of work to support external plugins defining new options:

Use cases:

- [ ]  This should include the options contributed by plugins:
  ```bash
  prettier --plugin @prettier/plugin-python --help
  ```

- [ ] CLI should work:
   ```bash
   prettier --plugin my-cool-prettier-plugin --my-custom-option
   ```

- [ ]  Remove unknown option warning for options contributed by external plugins.

   ```js
   prettier.format('def foo(): pass', {
     parser: 'python',
     plugins: ['@prettier/plugin-python'],
     pythonVersion: '3.1.4'
   })
  ● Validation Warning:

    Unknown option "pythonVersion" with value "3.1.4" was found.
    This is probably a typing mistake. Fixing it will remove this message.
  ```

### PR DESCRIPTION
**Pull request by ikatyang**
Fixes #3615

- ~~a lot of refactoring (actually most of them are just moved/extracted).~~
- remove `PRETTIER_LOG_LEVEL` environment variable, the logger is not global now.
- add a `showInternal` option for `supportInfo`, which is a filter for `cliSomething` used in CLI.
  - centralize infos
- support usage `prettier --plugin foo-plugin --help`.
- support detailed usage `prettier --plugin foo-plugin --help foo-option`.
- support external options from CLI `prettier --plugin foo-plugin --foo-option`.
- support external options from config file.
  - .prettierrc: `{ "plugins": ["foo-plugin"], "fooOption": "bar" }`
- **does not** support external plugins from config file but want to pass its options to CLI, vice versa.
  - .prettierrc: `{ "plugins": ["foo-plugin"] }` + CLI: `prettier --foo-option`

### CODE REVIEW

File: tests_integration/plugins/options/plugin.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    "foo-parser": {
+      parse: text => ({ text }),
+      astFormat: "foo-ast"
+    }
+  },
+  printers: {
+    "foo-ast": {
+      print: (path, options) =>
+        options.fooOption ? `foo:${options.fooOption}` : path.getValue().text,
+      options: {
```
Comment (azz) [inline tests_integration/plugins/options/plugin.js]: I'm not sure if plugins should belong to the printer or just the plugin. There could be options associated with parsers, too.

File: src/cli/context.js [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+const optionsNormalizer = require("../main/options-normalizer");
+const thirdParty = require("../common/third-party");
+const getSupportInfo = require("../common/support").getSupportInfo;
+const commonUtil = require("../common/util");
+
+const OPTION_USAGE_THRESHOLD = 25;
+const CHOICE_USAGE_MARGIN = 3;
+const CHOICE_USAGE_INDENTATION = 2;
+
+class Context {
```
Comment (azz) [inline src/cli/context.js]: I'm not a fan of a context god class. Would be much nicer if it was a simple `cliContext` object that was passed to all functions.

File: tests_integration/plugins/options/plugin.js (hunk 1 above)
Comment (ikatyang) [reply tests_integration/plugins/options/plugin.js]: IMO, options should be plugin-level, but that's the current behavior. It seems we're going to change it? then it should be in another PR.

Comment (azz) [reply tests_integration/plugins/options/plugin.js]: 👍  Happy for it to be a separate PR

**Comment by azz** [comment]
That looks a _lot_ better to me.

**Comment by ikatyang** [comment]
Any further comments?

**Pull request merged**